module/zfs/dmu_tx.c

```diff
@@ -1249,10 +1249,13 @@ dmu_tx_unassign(dmu_tx_t *tx)
 	 * Walk the transaction's hold list, removing the hold on the
 	 * associated dnode, and notifying waiters if the refcount drops to 0.
 	 */
-	for (txh = list_head(&tx->tx_holds); txh != tx->tx_needassign_txh;
+	for (txh = list_head(&tx->tx_holds); txh;
 	    txh = list_next(&tx->tx_holds, txh)) {
```

Comment:
I'd suggest this as a simpler version of the proposed fix.  The issue is that if `txh` is ever NULL then we'll NULL deref in `dnode_t *dn = txh->txh_dnode;` so let's do the check in the `for` loop to make it clear.

```
	for (txh = list_head(&tx->tx_holds);
	    txh && txh != tx->tx_needassign_txh;
  	    txh = list_next(&tx->tx_holds, txh)) {
```